internal/bminventory/inventory.go

```diff
@@ -2596,15 +2596,26 @@ func (b *bareMetalInventory) updateNonDhcpNetworkParams(updates map[string]inter
 	apiVip := cluster.APIVip
 	ingressVip := cluster.IngressVip
 
-	// We are checking if the cluster is requested to be a dual-stack cluster based on the Cluster
-	// Networks provided.
-	//
-	// TODO(mko) As updateNonDhcpNetworkParams is called before updateNetworks, this check looks
-	//           at the already configured networks and not the ones inside V2UpdateClusterParams.
-	//           An extension would be to check if V2UpdateClusterParams configures any of those
-	//           and if that's the case, instead of GetConfiguredAddressFamilies(cluster) we
-	//           should call something like GetRequestedAddressFamilies(cluster).
-	reqV4, reqV6, _ := network.GetConfiguredAddressFamilies(cluster)
+	// In order to check if the cluster is dual-stack or single-stack we are building a structure
+	// that is a merge of the current cluster configuration and new configuration coming from
+	// V2UpdateClusterParams. This ensures that the reqDualStack flag reflects a desired state and
+	// not the stale one.
+	// We need to do it because updateNonDhcpNetworkParams is called before updateNetworks, so
+	// inside this function here cluster object has still values before applying requested changes.
+	targetConfiguration := common.Cluster{}
+	targetConfiguration.ClusterNetworks = cluster.ClusterNetworks
+	targetConfiguration.ServiceNetworks = cluster.ServiceNetworks
+	targetConfiguration.MachineNetworks = cluster.MachineNetworks
+	if params.ClusterUpdateParams.ClusterNetworks != nil {
+		targetConfiguration.ClusterNetworks = params.ClusterUpdateParams.ClusterNetworks
+	}
+	if params.ClusterUpdateParams.ServiceNetworks != nil {
+		targetConfiguration.ServiceNetworks = params.ClusterUpdateParams.ServiceNetworks
+	}
+	if params.ClusterUpdateParams.MachineNetworks != nil {
+		targetConfiguration.MachineNetworks = params.ClusterUpdateParams.MachineNetworks
+	}
+	reqV4, reqV6, _ := network.GetConfiguredAddressFamilies(&targetConfiguration)
```

Comment:
Why error is not handled? Maybe it's a good chance to fix that as well?

Reply:
The flow of network updates is as follows
1) `updateNetworkParams`
2a)`updateDhcpNetworkParams` or
2b) `updateNonDhcpNetworkParams` or
2c)`setCommonUserNetworkManagedParams`
3) `updateNetworks`

This call here happens inside (2b). Inside (3) we have the following

```
if err = network.VerifyClusterOrServiceCIDR
[...]
if err = network.VerifyMachineCIDR
```
(which both are syntax sugar over `verifySubnetCIDR()` which amongst many other things, checks for `net.ParseCIDR` errors.

This flow allows us to keep some of the actions separate, as moving this handler to the place that you linked would mean that
* to be consistent, we should add an equivalent check somewhere in (2a)
* the check in (3) is redundant

Mostly because of the check in (3), I'm not very eager to also add the same in (2b). The current logic if you indeed have a "wrong" network in the payload would be
* the code you mentioned passes and marks cluster as non-dual-stack (because the unhandled error will return `(false, false`)
* no validation gets overriden because cluster is single-stack
* `updateNetworks` (which is called always) throws an error coming from `net.ParseCIDR` 

Let me know what you think, but I'm not sure if we need to handle the same error in multiple places

Reply:
My concern is what will happen if someone will change the implementation of `GetConfiguredAddressFamilies` and it will actually return an error that need to get back to the user, but maybe i'm missing something. 
I prefer not to count on the implementation details of this function in the main flow. 
But maybe i'm missing something.

Reply:
Fair point, I'll change it right away

Reply:
Changed. In fact we already have a wrapper function that covers error handling, so I'll just use that one instead